08-IO/Party.hs

```diff
@@ -0,0 +1,178 @@
+these to decide on the guest list for the next level up, and so on. So
+we need to write a combining function
+combineGLs :: Employee -> [GuestList] -> GuestList
+which takes an employee (the boss of some division) and the optimal
+guest list for each subdivision under him, and somehow combines
+this information to compute the best guest list for the entire division.
+However, this obvious first attempt fails! The problem is that we
+don’t get enough information from the recursive calls. If the best
+guest list for some subtree involves inviting that subtree’s boss, then
+we are stuck, since we might want to consider inviting the boss of the
+entire tree—in which case we don’t want to invite any of the subtree
+bosses (since they wouldn’t have any fun anyway). But we might be
+able to do better than just taking the best possible guest list for each
+subtree and then excluding their bosses.
+The solution is to generalize the recursion to compute more information, in such a way that we can actually make the recursive step.
+In particular, instead of just computing the best guest list for a given
+tree, we will compute two guest lists:
+1. the best possible guest list we can create if we invite the boss (that
+is, the Employee at the root of the tree); and
+2. the best possible guest list we can create if we don’t invite the boss.
+It turns out that this gives us enough information at each step to
+compute the optimal two guest lists for the next level up.
+-}
+
+{-
+Exercise 3
+Write a function
+nextLevel :: Employee -> [(GuestList, GuestList)]
+-> (GuestList, GuestList)
+which takes two arguments. The first is the “boss” of the current subtree (let’s call him Bob). The second argument is a list of the results
+for each subtree under Bob. Each result is a pair of GuestLists: the
+first GuestList in the pair is the best possible guest list with the boss
+of that subtree; the second is the best possible guest list without the
+boss of that subtree. nextLevel should then compute the overall best
+guest list that includes Bob, and the overall best guest list that doesn’t
+include Bob.
+-}
+nextLevel :: Employee -> [(GuestList, GuestList)] -> (GuestList, GuestList)
+nextLevel boss bestRes = first (glCons boss) (swap $ mconcat bestRes)
+
+{-
+Exercise 4
+Finally, put all of this together to define
+maxFun :: Tree Employee -> GuestList
+which takes a company hierarchy as input and outputs a fun-maximizing
+guest list. You can test your function on testCompany, provided in
+Employee.hs.
+The whole company
+Of course, the actual tree of employees in your company is much
+larger! We have provided you with a file, company.txt, containing
+the entire hierarchy for your company. The contents of this file were
+created by calling the show function on a Tree Employee,
+3
+so you can convert it back into a Tree Employee using the read function
+-}
+--Exercise 4
+maxFun :: Tree Employee -> GuestList
+maxFun = uncurry moreFun . treeFold nextLevel mempty
```

Comment:
Does this mempty work here? I'd expect that to be no longer needed now that after foldTree is simplified..

Reply:
@jhrcek Nope, it will no longer work. I have removed it.